### DISCUSSION TITLE
Choosing hyperperameters for minimum feature size constraint functions for Adjoint Optimization

### DISCUSSION THREAD

**Discussion by satadrudas** [NanoComp/meep#2689] [Q&A]
Hi,

I was going through the [Adjoint Solver tutorial with minimum feature size](https://meep.readthedocs.io/en/latest/Python_Tutorials/Adjoint_Solver/) in the MEEP documentation. Can someone please explain to me what does the parameters `a1`, `b1` and `c0` represent?

From the paper [Minimum length scale in topology optimization by geometric constraints](https://www.sciencedirect.com/science/article/pii/S0045782515001693), I got to know that `c0` is the decay rate of the Indicator functions. `c0` is used as an argument for `constraint_void()`  and `constraint_solid()`  in the `glc()` function.  In the [source code of `constraint_void()`  and `constraint_solid()` ](https://github.com/NanoComp/meep/blob/master/python/adjoint/filters.py) it is  mentioned that the the `c0` value lies withing the range (1e0 - 1e8). However, in the tutorial, `c0 = 1e7 * (filter_radius * 1 / resolution) ** 4
` which is equal to 0.0008100000000000001 since `filter_radius = 0.15` and `resolution = 50`. 

So why is the  value of `c0` not in the range (1e0 - 1e8) as mentioned in the source code?

Also in the paper [Phase-InjectedTopologyOptimizationfor ScalableandInterferometricallyRobustPhotonicIntegratedCircuits](https://pubs.acs.org/doi/10.1021/acsphotonics.2c01016#:~:text=This%20new%20approach%2C%20which%20we,existing%20adjoint-based%20design%20frameworks.), `c=8500`. How do we choose a good value for `c0` ? Trial and error ?

**Comment by smartalecH** [comment]
Trial and error, like most TO hyperparameters. 

Not super helpful, I know. But like machine learning, there's a lot of art to this. But unlike machine learning, there isn't enough investment in the tooling or methods to better democratize the process (or better yet, automate it altogether). 

I typically do quite a bit of hyperparameter optimization for each design problem.

**Reply by satadrudas** [reply]
Hi, thanks for your answer. 

Also, I did understand the purpose of `c0`, but could you also please teel me the role of `a1` and `b1` for completeness?
